### PR TITLE
correct pointing in "@projectname --project" testing

### PR DESCRIPTION
**Pull request by ngam**
- ran into this while customizing conda-forge build of julia
- with @mkitti

Originally, the test as written would always default to home. However, if someone sets a different `JULIA_DEPOT_PATH`, then the test would fail. This slight edit ensures it passes. 

xref: https://github.com/conda-forge/julia-feedstock/pull/157

A simple test to confirm the problem:

```bash
JULIA_DEPOT_PATH="/some/path/of/choice:$JULIA_DEPOT_PATH" julia -E 'Base.runtests(["cmdlineargs"]; ncores=ceil(Int, Sys.CPU_THREADS))'
```

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Won't this error if the user doesn't have `JULIA_DEPOT_PATH` in their environment?

**Comment by ngam** [comment]
> Won't this error if the user doesn't have `JULIA_DEPOT_PATH` in their environment?

I believe `JULIA_DEPOT_PATH` takes on some default values: https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_DEPOT_PATH

**Comment by ngam** [comment]
> Won't this error if the user doesn't have `JULIA_DEPOT_PATH` in their environment?

Ah, you're right. I thought it would populate it automatically. What would you recommend here, then? @DilumAluthge

File: test/cmdlineargs.jl
Comment (DilumAluthge) [inline test/cmdlineargs.jl:142]:
```suggestion
        let
            env = Dict{String, String}()
            env["JULIA_PROJECT"] = "@foo"
            env["HOME"] = homedir()
            if haskey(ENV, "JULIA_DEPOT_PATH")
                env["JULIA_DEPOT_PATH"] = ENV["JULIA_DEPOT_PATH"]
            end
            @test expanded == readchomp(addenv(`$exename -e 'println(Base.active_project())'`, env))
        end
```

Comment (DilumAluthge) [reply test/cmdlineargs.jl:142]: And similarly for any other tests that need this fix.

Comment (ngam) [reply test/cmdlineargs.jl:142]: Hm. I see what you did there. How about a shortcut? `Base.DEPOT_PATH[1]` will always yield the set `DEPOT_PATH` or will fall back on the default `~/.julia`

Comment (DilumAluthge) [reply test/cmdlineargs.jl:142]: That should work.

Comment (ngam) [reply test/cmdlineargs.jl:142]: I will commit your suggestion, but let's consider a shortcut.

Comment (DilumAluthge) [inline test/cmdlineargs.jl]:
```suggestion
        @test expanded == readchomp(addenv(`$exename -e 'println(Base.active_project())'`, "JULIA_PROJECT" => "@foo", "HOME" => homedir(), "JULIA_DEPOT_PATH" => "$(Base.DEPOT_PATH[1])"))
        
        
```

Comment (ngam) [reply test/cmdlineargs.jl]: Oops. Let me fix the mess I created :D

Comment (ngam) [reply test/cmdlineargs.jl]: THANKS A LOT for catching this!

**Comment by DilumAluthge** [comment]
1. Can you delete lines 143 and 144 from ` test/cmdlineargs.jl`?
2. Can you squash?

File: test/cmdlineargs.jl
Comment (DilumAluthge) [inline test/cmdlineargs.jl:142]: I think we should actually just delete the `JULIA_DEPOT_PATH` environment variable.

```suggestion
        @test expanded == readchomp(addenv(`$exename -e 'println(Base.active_project())'`, "JULIA_PROJECT" => "@foo", "HOME" => homedir(), "JULIA_DEPOT_PATH" => nothing))
```

Comment (ngam) [reply test/cmdlineargs.jl:142]: Why? That won't work, because it would essentially point it to a folder named nothing in the current working directory. At least that's how it plays out on my machine rn...

Comment (DilumAluthge) [reply test/cmdlineargs.jl:142]: `"JULIA_DEPOT_PATH" => nothing` is equivalent to doing `unset JULIA_DEPOT_PATH` in Bash or doing `delete!(ENV, "JULIA_DEPOT_PATH") in Julia.

Comment (DilumAluthge) [reply test/cmdlineargs.jl:142]: Hmmmm. At least, that is what should happen. It doesn't seem to be working correctly.

Comment (ngam) [reply test/cmdlineargs.jl:142]: But we want to make sure that that the created env is in wherever the user is setting it... removing it will take us back to square one, if not worse. I am still confused how this will resolve the issue.

Comment (DilumAluthge) [reply test/cmdlineargs.jl:142]: I guess it works for `withenv`, but not for `addenv` or `setenv`.

File: test/cmdlineargs.jl
Comment (DilumAluthge) [inline test/cmdlineargs.jl:142]:
```suggestion
        env = copy(ENV)
        env["JULIA_PROJECT"] = "@foo"
        env["HOME"] = homedir()
        delete!(env, "JULIA_DEPOT_PATH")
        @test expanded == readchomp(addenv(`$exename -e 'println(Base.active_project())'`, env))
```

Comment (DilumAluthge) [reply test/cmdlineargs.jl:142]: @ngam Try this out locally.

**Comment by ngam** [comment]
> * Can you delete lines 143 and 144 from ` test/cmdlineargs.jl`?
> * Can you squash?

I delete the empty lines. Should I still squash now or is it okay to squash upon (potential) merge?

**Comment by DilumAluthge** [comment]
Let's wait to squash until we're happy with the implementation.

File: test/cmdlineargs.jl
Comment (ngam) [reply test/cmdlineargs.jl:142]: ✅

Comment (ngam) [reply test/cmdlineargs.jl:142]: I like your way of doing this because it makes it much clearer, as opposed to the previous implementation.

Comment (ngam) [reply test/cmdlineargs.jl:142]: Let me think a little about this and I will revisit. We are trying to figure out a decent solution for the conda-forge's build of julia (basically conda a whole package manager, and julia has effectively another package manager as well at this point, and so we want to make sure we isolate them carefully so that people can get the full use of julia from conda-forge)

Comment (DilumAluthge) [reply test/cmdlineargs.jl:142]: I think that explicitly deleting the `JULIA_DEPOT_PATH` environment variable (by doing `delete!(env, "JULIA_DEPOT_PATH")`) is the most correct solution here, because it makes explicit what we are trying to accomplish.

**Comment by DilumAluthge** [comment]
I would say:
1. Test my suggestion (https://github.com/JuliaLang/julia/pull/43560#discussion_r775327354) locally to make sure it works for your use case.
2. Accept the suggestion.
3. Rebase and squash.
4. Wait for CI to pass.
5. Ping me once CI is passing, and I can merge.

File: test/cmdlineargs.jl
Comment (DilumAluthge) [inline test/cmdlineargs.jl:142]: Since we are doing `env = copy(ENV)` and then deleting keys from `env`, we should use `setenv` instead of `addenv`.

```suggestion
        env = copy(ENV)
        env["JULIA_PROJECT"] = "@foo"
        env["HOME"] = homedir()
        delete!(env, "JULIA_DEPOT_PATH")
        @test expanded == readchomp(setenv(`$exename -e 'println(Base.active_project())'`, env))
```

**Review by DilumAluthge — CHANGES_REQUESTED** [review]

File: test/cmdlineargs.jl
Comment (DilumAluthge) [reply test/cmdlineargs.jl:142]: As I mentioned in a different thread:

> I think that explicitly deleting the JULIA_DEPOT_PATH environment variable (by doing delete!(env, "JULIA_DEPOT_PATH")) is the most correct solution here, because it makes explicit what we are trying to accomplish.

**Comment by mkitti** [comment]
Abort this pull request. Do not merge.

@ngam The patch is not needed. As I mentioned in https://github.com/conda-forge/julia-feedstock/pull/157#issuecomment-1001279748, I thought this might have already been fixed upstream. It has. It already have been fixed by @DilumAluthge in https://github.com/JuliaLang/julia/pull/42358 .

@KristofferC could we backport #42358 to 1.7 branch?

**Comment by mkitti** [comment]
The [originating error](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=432811&view=logs&jobId=656edd35-690f-5c53-9ba3-09c10d0bea97&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d) on 1.7.1 is as follows:

```
Test Failed at $PREFIX/share/julia/test/cmdlineargs.jl:122
  Expression: expanded == readchomp(setenv(`$exename -e 'println(Base.active_project())'`, "JULIA_PROJECT" => "@foo", "HOME" => homedir()))
   Evaluated: "$PREFIX/share/julia/site/environments/foo/Project.toml" == "/home/conda/.julia/environments/foo/Project.toml"
```

This error occurs when testing 1.7.1 since that still uses `setenv`.

In #42358 (which is merged on Julia master slated for Julia 1.8), this was changed to `addenv`, which allows `JULIA_DEPOT_PATH` to be inherited by the new process. Therefore, we do not need to explicitly set `JULIA_DEPOT_PATH` due to the use of  `addenv`. Additionally, that PR explicitly tested the use of `JULIA_DEPOT_PATH`.

The problem that we are trying to address here does not exist on Julia master.

**Comment by ngam** [comment]
> The [originating error](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=432811&view=logs&jobId=656edd35-690f-5c53-9ba3-09c10d0bea97&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d) on 1.7.1 is as follows:
> 
> ```
> Test Failed at $PREFIX/share/julia/test/cmdlineargs.jl:122
>   Expression: expanded == readchomp(setenv(`$exename -e 'println(Base.active_project())'`, "JULIA_PROJECT" => "@foo", "HOME" => homedir()))
>    Evaluated: "$PREFIX/share/julia/site/environments/foo/Project.toml" == "/home/conda/.julia/environments/foo/Project.toml"
> ```
> 
> This error occurs when testing 1.7.1 since that still uses `setenv`.
> 
> In #42358 (which is merged on Julia master slated for Julia 1.8), this was changed to `addenv`, which allows `JULIA_DEPOT_PATH` to be inherited by the new process. Therefore, we do not need to explicitly set `JULIA_DEPOT_PATH` due to the use of `addenv`. Additionally, that PR explicitly tested the use of `JULIA_DEPOT_PATH`.
> 
> The problem that we are trying to address here does not exist on Julia master.

It seems to me that PR #42358 only addresses this insofar as `Base.runtests(["all"])` is called.

**Comment by DilumAluthge** [comment]
Oh I'd totally forgotten about https://github.com/JuliaLang/julia/pull/42358.

Yeah https://github.com/JuliaLang/julia/pull/42358 solved this problem.

**Comment by DilumAluthge** [comment]
> It seems to me that PR #42358 only addresses this insofar as `Base.runtests(["all"])` is called.

Hmmm. https://github.com/JuliaLang/julia/pull/42358 should solve this for any invocation of `Base.runtests`, regardless of the arguments passed to `Base.runtests`.

**Comment by DilumAluthge** [comment]
I think the issue here is that https://github.com/JuliaLang/julia/pull/42358 has not yet been backported to 1.6 and 1.7.

**Comment by ngam** [comment]
> > It seems to me that PR #42358 only addresses this insofar as `Base.runtests(["all"])` is called.
> 
> Hmmm. #42358 should solve this for any invocation of `Base.runtests`, regardless of the arguments passed to `Base.runtests`.

Sorry, I misread that. Thanks!

**Comment by ngam** [comment]
Btw @DilumAluthge --- I did my undergrad degree at Brown, _in BME_. Small world.

**Comment by ngam** [comment]
> Btw @DilumAluthge --- I did my undergrad degree at Brown, _in BME_. Small world.

@DilumAluthge okay... now... I think I remember you. I believe we took APMA classes together! I graduated in 2015

**Pull request closed without merging**
